workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/src/Generation/Model/Details/DetailsStore.cs

```diff
@@ -0,0 +1,152 @@
+
+        private Dictionary<Identifier, bool> blittableMap = new Dictionary<Identifier, bool>();
+
+        private readonly SchemaBundle bundle;
+
+        public DetailsStore(SchemaBundle bundle)
+        {
+            this.bundle = bundle;
+
+            PopulateBlittableMaps();
+            BlittableMap = ImmutableHashSet.CreateRange(blittableMap.Where(kv => kv.Value).Select(kv => kv.Key));
+
+            var enums = bundle.BundleContents.EnumDefinitions
+                .Select(enumm => (enumm.EnumIdentifier, new UnityEnumDetails(enumm)))
+                .ToDictionary(tuple => tuple.Item1, tuple => tuple.Item2);
+
+            Enums = new ReadOnlyDictionary<Identifier, UnityEnumDetails>(enums);
+
+            var types = bundle.BundleContents.TypeDefinitions
+                .Select(type => (type.Identifier, new UnityTypeDetails(type)))
+                .ToDictionary(tuple => tuple.Item1, tuple => tuple.Item2);
+
+            Types = new ReadOnlyDictionary<Identifier, UnityTypeDetails>(types);
+
+            var components = bundle.BundleContents.ComponentDefinitions
+                .Select(component => (component.Identifier, new UnityComponentDetails(component, this)))
+                .ToDictionary(tuple => tuple.Item1, tuple => tuple.Item2);
+
+            Components = new ReadOnlyDictionary<Identifier, UnityComponentDetails>(components);
+
+            SchemaFiles = bundle.SourceMap.SourceReferences.Values
+                .Select(sourceRef => sourceRef.FilePath)
+                .Distinct()
+                .ToList()
+                .AsReadOnly();
+
+            foreach (var kv in Types)
+            {
+                kv.Value.PopulateFields(this);
+                kv.Value.PopulateChildren(this);
+            }
+
+            foreach (var kv in Components)
+            {
+                kv.Value.PopulateFields(this);
+            }
+        }
+
+        public HashSet<Identifier> GetNestedTypes(Identifier identifier)
+        {
+            return Types.Select(t => t.Key).Where(t => IsIdentifierChild(identifier, t))
+                .Concat(Enums.Select(t => t.Key).Where(t => IsIdentifierChild(identifier, t)))
+                .ToHashSet();
+        }
+
+        private bool IsIdentifierChild(Identifier parent, Identifier potentialChild)
+        {
+            return potentialChild.QualifiedName.StartsWith(parent.QualifiedName)
+                && potentialChild.Path.Count == parent.Path.Count + 1;
+        }
+
+        private void PopulateBlittableMaps()
+        {
+            // Populate all enums.
+            foreach (var enumm in bundle.BundleContents.EnumDefinitions)
+            {
+                blittableMap.Add(enumm.EnumIdentifier, true);
+            }
+
+            var typesToTraverse = new Queue<TypeDefinitionRaw>(bundle.BundleContents.TypeDefinitions);
+
+            while (typesToTraverse.Count > 0)
```

Comment:
Is there a chance we would go into infinite loop here if two types are referencing each other?

Reply:
No as the only way that can happen is through a reference type which marks it as non-blittable immediately. (I believe)